google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,129 @@
+using StreamInitializer = std::function<future<StatusOr<
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>>(
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>)>;
+
+using AsyncSleeper = std::function<future<void>(std::chrono::milliseconds)>;
+
+using RetryPolicyFactory = std::function<std::unique_ptr<RetryPolicy>()>;
+
+/**
+ * The purpose of this class is to encapsulate the retry logic of a generic
+ * bidirectional asynchronous RPC stream.
+ */
+template <typename RequestType, typename ResponseType>
+class ResumableAsyncStreamingReadWriteRpc {
+ public:
+  virtual ~ResumableAsyncStreamingReadWriteRpc() = default;
+
+  /**
+   * Start the streaming RPC.
+   *
+   * The returned future will be returned with a status
+   * when this stream will not be resumed or when the user calls `Finish()`. In
+   * the case that there are no errors from `Start`ing the stream and on the
+   * latest `Read` and `Write` calls if present, this will return an ok
+   * `Status`.
+   */
+  virtual future<Status> Start() = 0;
+
+  /**
+   * Read one response from the streaming RPC.
+   *
+   * @note Only **one** `Read()` operation may be pending at a time. The
+   * application is responsible for waiting until any previous `Read()`
+   * operations have completed before calling `Read()` again.
+   *
+   * Whether `Read()` can be called before a `Write()` operation is specified by
+   * each service and RPC. Most services require at least one `Write()` call
+   * before calling `Read()`. Many services may return more than one response
+   * for a single `Write()` request.  Each service and RPC specifies how to
+   * discover if more responses will be forthcoming.
+   *
+   * If the `optional<>` is engaged, a successful `ResponseType` is
+   * returned. If it is not engaged, the call failed, but the user may call
+   * `Read` again unless `Start` had finished with a permanent error `Status`.
+   */
+  virtual future<absl::optional<ResponseType>> Read() = 0;
+
+  /**
+   * Write one request to the streaming RPC.
+   *
+   * @note Only **one** `Write()` operation may be pending at a time. The
+   * application is responsible for waiting until any previous `Write()`
+   * operations have completed before calling `Write()` again.
+   *
+   * Whether `Write()` can be called before waiting for a matching `Read()`
+   * operation is specified by each service and RPC. Many services tolerate
+   * multiple `Write()` calls before performing or at least receiving a `Read()`
+   * response.
+   *
+   * If `true` is returned, the call was successful. If `false` is returned,
+   * the call failed, but the user may call `Write` again unless `Start` had
+   * finished with a permanent error `Status`.
+   */
+  virtual future<bool> Write(RequestType const&, grpc::WriteOptions) = 0;
+
+  /**
+   * Finishes the streaming RPC.
+   *
+   * This will cause any outstanding `Read` or `Write` to fail.
```

Comment:
Note that this is not the behavior in gRPC:

https://grpc.github.io/grpc/cpp/classgrpc_1_1internal_1_1_client_async_streaming_interface.html#a0c4fd4be2228fa196ae446502828e23a

In particular, calling `Finish()` for the underlying stream is only possible if:

> there are no more messages to be received from the server (this can be known implicitly by the calling code, or explicitly from an earlier call to [AsyncReaderInterface::Read](https://grpc.github.io/grpc/cpp/classgrpc_1_1internal_1_1_async_reader_interface.html#adf87c602036d69158950b7299d0aae70) that yielded a failed result, e.g. cq->Next(&read_tag, &ok) filled in 'ok' with 'false').

Is your intention is to call `Finish()` (in the underlying stream) while a `Read()` call is outstanding?  Then you will also need to (a) call `Cancel()` to terminate the `Read()` call, then (b) drain all the `Read()` calls that return `true`, and finally (c) call `Finish()` after the first `Read()` that returns `false`.

I am not sure your comments / API proposal are necessarily bad or not.  At this point I am just asking what you want the API to do.

Reply:
This is the intention yes, which could be more clear in documentation. It is to abstract away handling of how Read() and Write() calls to this object correspond to Read() and Write() calls on the underlying stream object, which may change over time as streams fail.

The intention of this is that, by the time Finish() returns, the future returned from `Start` will have the final status (if it doesn't already due to a permanent error), any outstanding Read() and Write() futures from this object will be failed, and any future calls to Read() or Write() will fail immediately.

This is largely to encapsulate the retry logic needed for classes like StreamingSubscriptionBatchSource, of which there will be 4 in Pub/Sub Lite code, to prevent needing to write complicated retry logic multiple times.

 https://github.com/googleapis/google-cloud-cpp/blob/c9f7f317a245fbeac2af634953ccc3d6eb6e639b/google/cloud/pubsub/internal/streaming_subscription_batch_source.h#L63